### ISSUE TITLE
Emscripten Broken 

### ISSUE THREAD

**Issue by LDighera** [emscripten-core/emscripten#4306]
In order to build the JS version of csdr, as is used by openwebrx,
emscripten is required but this is currently broken in sid. Blocking the
ITP by the relevant emscripten bugs.

This packaging work will take place within the Debian Hamradio
Maintainers team debian-hams@lists.debian.org and is being done as
this is a dependency for OpenWebRX.

As posted by:  "Iain R. Learmonth" irl@debian.org

**Comment by kripken** [comment]
Sounds like that's a Debian packaging issue - is there anything we can do to help here in upstream?

**Comment by LDighera** [comment]
Kripken,

I received the message below from András Retzler:

> To: Larry Dighera <LDighera@att.net>
> Subject: Re: (fwd) Bug#821115: need emscripten for csdr
> From: Retzler András <retzlerandras@gmail.com>
> Date: Sun, 15 May 2016 11:11:48 +0200
> 
> Iain contacted me about openwebrx and csdr packaging to Debian, and soon
> I'm going to do what he asked for. (See GitHub
> <https://github.com/simonyiszk/csdr/issues/created_by/irl>.)
> 
> In order to build the JS version of csdr, as is used by openwebrx,
> emscripten is required but this is currently broken in sid. Blocking the
> ITP by the relevant emscripten bugs.
> 
> As far as I understand, this issue is with Emscripten, not csdr. I am using
> an Ubuntu-based distribution for development and the emscripten in the
> Ubuntu repositories is okay, I can compile csdr (sdr.js) with it.
> 
> If you do not want to build sdr.js yourself, you can find a compiled
> version of sdr.js in the OpenWebRX repository:
> https://raw.githubusercontent.com/simonyiszk/openwebrx/master/htdocs/sdr.js
> 
> Andras, HA7ILM

**Comment by kripken** [comment]
I may be misunderstanding, but it mentions "relevant emscripten bugs" - are there bugs we should fix in upstream? In those links I don't see mention of them, so I guess there aren't?

**Comment by irl** [comment]
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=775147
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=809997

These were the relevant bugs.

**Comment by irl** [comment]
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=775147#10

This is the primary problem. emscripten is not friendly to distribution packagers as it requires this custom LLVM magic.

**Comment by juj** [comment]
> This is the primary problem. emscripten is not friendly to distribution packagers as it requires this custom LLVM magic.

This somehow builds on the assumption that only one instance of LLVM shall exist on a given system. Emscripten uses its own branch of LLVM, and does so internally, so distribution packagers should bundle Emscripten's branch of LLVM into the package, in a way that it is stored internally in Emscripten installation tree and not "advertised" as any kind of "system default" LLVM when compiling anything else. I am not familiar with linux packaging architectures, but surely it should be possible to have packages contain their own internal executables that aren't exported system-wide(?)

Generally though we don't recommend developers to obtain Emscripten via package managers, but bootstrapping via the emsdk is preferred, because that allows updating between latest stable and development branches, so that it's easy to update to newest compiler versions immediately when they become available, without having to wait for a distribution update cycle. If e.g. a user reports a bug and it is fixed, emsdk allows an easy zero-day update to the latest compiler branch to try out the fix.

**Comment by irl** [comment]
Oh dear. :/

> distribution packagers should bundle Emscripten's branch of LLVM into the package

No. This is a violation of Debian Policy [§4.13](https://www.debian.org/doc/debian-policy/ch-source.html#s-embeddedfiles). Debian Policy [footnote 30](https://www.debian.org/doc/debian-policy/footnotes.html#f30) expands on the reasoning for this: "Having multiple copies of the same code in Debian is inefficient, often creates either static linking or shared library conflicts, and, most importantly, increases the difficulty of handling security vulnerabilities in the duplicated code. "

> Generally though we don't recommend developers to obtain Emscripten via package managers

This is perfectly fine for *developers*. Debian is not packaging primarily for developers though, but for users. The software I would like to package currently cannot be shipped in Debian because Debian cannot build it. We can't build it without packaging the toolchain used to build it and that toolchain isn't packagable as it currently stands.

While emscripten is currently packaged in Debian (by @sylvestre), I believe that if this will continue to be the way in which development continues (i.e. not being able to be used with vanilla upstream LLVM, not pushing patches upstream) then emscripten does not belong in Debian and should be removed and relegated to the pool of esoteric "hobby project"-grade software that cannot easily be maintained and is not suitable for production use.

**Comment by sylvestre** [comment]
Well, I stopped maintaining it for this reason. emscripten (an old version) is only available in Debian Unstable.

**Comment by juj** [comment]
> "Having multiple copies of the same code in Debian ...

That sounds ok, although it could be argued here that it is not the same code, since if it was the same code, it would have had the same functionality as well. But perhaps the "same code" is interpreted loosely there. If we could choose, we would add our backend to LLVM upstream so that it builds with it as-is, but it's not something we could just go and dictate.

> ... then emscripten does not belong in Debian and should be removed and relegated to the pool of esoteric "hobby project"-grade software that cannot easily be maintained and is not suitable for production use.

Understand the sentiment, but I hope that is not a reason to call [sour grapes](https://en.wikipedia.org/wiki/The_Fox_and_the_Grapes#The_fable). It sounds like if the rules are so strict in Debian, Emscripten might not be suitable for production use **in Debian**.

We hope to see a way for Debian to improve here in the future. Perhaps a way to resolve is when the upstream LLVM WebAssembly backend module is ready and makes its way to Debian as well. Emscripten is targeting this backend, which is sometimes called the "vanilla LLVM wasm backend" in Emscripten community, but wasm hasn't yet shipped even in browsers yet, so that is too early. Perhaps a good time for Debian packagers might be to revisit the issue after LLVM can be built with wasm backend bundled. Although that would require that Debian would choose to enable the backend in its system compiler, as it is not a default needed feature when targeting native code.

**Comment by sylvestre** [comment]
> Perhaps a good time for Debian packagers might be to revisit the issue after LLVM can be built with wasm  backend bundled.

This has been my goal. I was hoping that the changes could be merged earlier in LLVM but I will wait...
Maintaining the LLVM fork in emscripten is not something that I want to do. It is too much work (I am already maintaining LLVM on the side and I don't want to sync the llvm upstream changes into emscripten).
This is also why we are building Rust using LLVM packages and not using llvm embedded sources (we even remove them from the tarball).

**Comment by irl** [comment]
Ok cool. So for now [packaging csdr](https://bugs.debian.org/821115) can continue to be blocked by [emscripten not being installable in sid](https://bugs.debian.org/809997) and we can see that we have a path towards that bug being resolved in the future by there being a future Debian package of LLVM that has the wasm backend enabled in it.

I'm not convinced this issue has a useful title or description, but this is useful information, so perhaps someone could create a new issue to register that requiring a custom set of LLVM sources is a bug and something we want to have fixed in the future? We could then have a Debian bug that is pointing towards that bug so that all the bugs make sense.
